particle-iot/device-os
# Increase the maximum size of event data

### Problem

This PR introduces the following changes:

- Increase the maximum size of payload data that can be sent or received in a cloud event as well as the maximum size of a cloud function's argument and a variable's value from 622 to 1024 bytes on Gen 3 platforms and from 622 to 864 bytes on Gen 2 platforms. On Gen 3 platforms, this change increases RAM usage by 1.4K.
- Extend the Hello message to notify the cloud of the additional session parameters: maximum supported size of a CoAP message, maximum supported firmware size, size of OTA update chunk, module version of the system firmware.
- Remove the events that have been superseded by the new Hello message fields (`spark/hardware/max_binary` and `spark/hardware/ota_chunk_size`).

### Steps to Test

- Flash the example application to a device and verify that it can successfully publish an event containing 1024 bytes of data.
- Verify that the device can receive an event containing 1024 bytes of data. For example:
```sh
cat /dev/urandom | tr -dc A-Za-z0-9 | head -c1024 > 1024.txt
particle publish event $(cat 1024.txt)
```
**Note:** This test requires changing the API limit for the maximum size of event data.
- Verify that the device's cloud function (`fn`) can be called with 1024 bytes of argument data:
```sh
particle function call my_device fn $(cat 1024.txt)
```
- Run integration tests.

### Example App

```cpp
#include "application.h"
#include "random.h"

namespace {

const auto FN_NAME = "fn";
const auto EVENT_NAME = "event";
const size_t EVENT_SIZE = 1024;

const SerialLogHandler logHandler(LOG_LEVEL_WARN, {
    { "app", LOG_LEVEL_ALL }
});

char buffer[EVENT_SIZE + 1] = {}; // +1 byte for null terminator
bool done = false;

int fn(const String& arg) {
    Log.info("Received function call; argument:");
    Log.print(arg);
    Log.printf("\r\n(%u bytes)\r\n", (unsigned)arg.length());
    return 0;
}

void event(const char* name, const char* data) {
    Log.info("Received event; name: %s; data:", name);
    Log.print(data);
    Log.printf("\r\n(%u bytes)\r\n", (unsigned)strlen(data));
}

} // namespace

SYSTEM_MODE(SEMI_AUTOMATIC)
SYSTEM_THREAD(ENABLED)

void setup() {
    Particle.function(FN_NAME, fn);
    Particle.subscribe(EVENT_NAME, event);
    Log.info("Connecting");
    Particle.connect();
}

void loop() {
    if (Particle.connected() && !done) {
        done = true;
        Log.info("Connected");
        delay(1000);
        Random().genBase32(buffer, EVENT_SIZE);
        Log.info("Publishing event; name: %s; data:", EVENT_NAME);
        Log.print(buffer);
        Log.printf("\r\n(%u bytes)\r\n", (unsigned)EVENT_SIZE);
        Particle.publish("event", buffer, WITH_ACK)
            .onSuccess([](bool) {
                Log.info("Successfully published event");
            })
            .onError([](const Error& e) {
                Log.info("Failed to publish event: %d", (int)e.type());
            });
    }
}
```

### References

- [ch64987]
- [ch64724]
